crates/rpc/rpc-builder/src/metrics.rs

```diff
@@ -247,3 +247,121 @@ struct RpcServerCallMetrics {
     /// Response for a single call
     time_seconds: Histogram,
 }
+
+/// A service that intercepts RPC calls and forwards pre-bedrock historical requests
+/// to a dedicated endpoint.
+#[derive(Debug, Clone)]
+pub struct HistoricalRpcService<S, P> {
+    /// The inner service that handles regular RPC requests
+    inner: S,
+    /// Client used to forward historical requests  
+    historical_client: Arc<HttpClient>,
+    /// Provider used to determine if a block is pre-bedrock
+    provider: Arc<P>,
+    /// Bedrock transition block number
+    bedrock_block: BlockNumber,
+}
+
+impl<S, P> RpcServiceT for HistoricalRpcService<S, P>
+where
+    S: RpcServiceT<MethodResponse = MethodResponse> + Send + Sync + Clone + 'static,
+    P: BlockReaderIdExt + Send + Sync + Clone + 'static,
+{
+    type MethodResponse = S::MethodResponse;
+    type NotificationResponse = S::NotificationResponse;
+    type BatchResponse = S::BatchResponse;
+
+    fn call<'a>(&self, req: Request<'a>) -> impl Future<Output = Self::MethodResponse> + Send + 'a {
+        let method = req.method.as_ref();
+        let inner_service = self.inner.clone();
+        let historical_client = self.historical_client.clone();
+        let provider = self.provider.clone(); // Fixed variable name
+        let bedrock_block = self.bedrock_block;
+
+        async move {
+            // Check if this method should be considered for historical forwarding
+            let is_historical_method = matches!(
+                method,
+                "eth_getBalance"
+                    | "eth_getStorageAt"
+                    | "eth_getCode"
+                    | "eth_call"
+                    | "eth_getTransactionCount"
+                    | "eth_getBlockByHash"
+                    | "eth_getBlockByNumber"
+                    | "eth_getTransactionByHash"
+                    | "eth_getTransactionReceipt"
+            );
```

Comment:
if we're doing specialized endpoint handling then we should do this as a match over the method, because now we're doing both